File: python/p0049_group_anagrams/solutions/brute_force_sorting.py
# Time complexity: O(n * k log k)
# Space complexity: O(n)
# where n = number of words, k = average length of a word

from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:

        anagram_map: dict = {}

        for word in strs:
            word_sorted = "".join(sorted(word))

            if word_sorted in anagram_map:
                anagram_map[word_sorted].append(word)
            else:
                anagram_map[word_sorted] = [word]

        anagrams_grouped = [group for group in anagram_map.values()]

        return anagrams_grouped

File: csharp/P0049_GroupAnagrams/Solutions/BruteForceSorting.cs
// Time complexity: O(n * k log k)
// Space complexity: O(n)
// where n = number of words, k = length of the longest word

namespace P0049_GroupAnagrams.Solutions.BruteForceSorting;

public class Solution
{
    public IList<IList<string>> GroupAnagrams(string[] strs)
    {
        // Build anagram map
        Dictionary<String, List<String>> anagramMap = new();

        foreach (String word in strs)
        {
            String wordSorted = new String(word.OrderBy(letter => letter).ToArray());

            if (!anagramMap.ContainsKey(wordSorted))
                anagramMap[wordSorted] = new List<String>();

            anagramMap[wordSorted].Add(word);
        }

        return anagramMap.Values.Cast<IList<String>>().ToList();
    }
}